ml_tools/feature_generator.py

```diff
@@ -0,0 +1,362 @@
+import pandas as pd
+from uszipcode import SearchEngine
+from abc import ABC, abstractmethod
+
+"""Module for generating, transforming and aggregating features of datasets.
+  
+The ml_tools module is a tool for transforming, aggregating and generating new features from existing features
+in input datasets. Users can use built in generators or define their own generating functions and have them
+incorporated into the API. 
+  
+  Typical usage examples:
+
+  feature_generator = GeneratorClass()
+  new_data = feature_generator.generate_feature(data)
+  
+  data_aggregator = AggregatorClass(data, relationship key)
+  aggregated_data = data_aggregator.aggregate()
+  
+  user_defined_generator = custom_generator(user_defined_function)
+"""
+
+
+class FeatureGenerator(ABC):
+    """Abstract base class for all feature generators and transformers
+
+    Attributes:
+        name: name of transformation or generator (string)
+        feature_type: type of feature being generated (string)
+    """
+
+    @property
+    def name(self):
+        raise NotImplementedError('Name property is not implemented')
+
+    @property
+    def feature_type(self):
+        raise NotImplementedError('feature_type property is not implemented')
+
+    @abstractmethod
+    def generate_feature(self, data, column=None, **kwargs):
+        """Abstract method for creating and transforming features
+
+        :param data: Pandas dataframe
+        :param column: Column on which to perform operation
+        :param kwargs: additional parameters for specific implementations
+        :return: Pandas dataframe with new or transformed features
+        """
+        pass
+
+
+class Hour(FeatureGenerator):
+    """Transforms datetime data to hour of the day"""
+
+    @property
+    def name(self):
+        return 'Hour'
+
+    @property
+    def feature_type(self):
+        return 'transformation'
+
+    @classmethod
+    def generate_feature(cls, data, column=None, **kwargs):
+        """Implements generate feature
+
+        Converts all datetime data to hours or, if given a column, converts a single column to hours
+
+        :param data: input dataframe
+        :param column: column with timestamp data
+        :param kwargs: ignored
+        :return: dataframe with timestamps converted to hours
+        """
+        if column is None:
+            object_cols = [col for col, col_type in data.dtypes.iteritems() if col_type == 'object']
+
+            for col in object_cols:
+                data[col] = pd.to_datetime(data[col], errors='ignore')
+
+            types = data.dtypes
+            ts_indices = [idx for idx, t in enumerate(list(types)) if t == "datetime64[ns]"]
+            for idx in ts_indices:
+                for index, time in data.iloc[:, idx].iteritems():
+                    data.iloc[:, idx][index] = time.hour
+            return data
+        else:
+            try:
```

Comment:
Do you want this try/except to be inside the for loop?  What is the condition raising the exception?